[[ Foundation ]] Improve checks for invalid ValueRefs.

runrevmark opened the pull request:
A new private debug assertion macro __MCAssertIsValue has been added. This
checks that the passed valueref has reference count > 0 and flags != -1.

Additionally, the value pool list link is now stored in the first pointer
sized field after the __MCValue header. This means that it is a lot less
likely that dangling references to valuerefs will break the value pool
free lists.

runrevmark committed:
[[ Foundation ]] Improve checks for invalid ValueRefs.

A new private debug assertion macro __MCAssertIsValue has been added. This
checks that the passed valueref has reference count > 0 and flags != -1.

Additionally, the value pool list link is now stored in the first pointer
sized field after the __MCValue header. This means that it is a lot less
likely that dangling references to valuerefs will break the value pool
free lists.

peter-b commented on libfoundation/src/foundation-private.h:
```diff
@@ -624,6 +624,9 @@ __MCAssertResolvedTypeInfo(MCTypeInfoRef x, bool (*p)(MCTypeInfoRef))
 #define __MCAssertValueType(x,T) MCAssert(MCValueGetTypeCode(x) == kMCValueTypeCode##T)
 
+// A valid ValueRef must have references > 0 and flags != -1
+#define __MCAssertIsValue(x)    MCAssert(((__MCValue *)x) -> flags != 0xffffffff && ((__MCValue *)x) -> references > 0);
```
`MCValue.flags` is `uint32_t`, so this should be `UINT32_MAX`

peter-b replied:
I think it might be worth making `__MCAssertIsValue(x)` also assert that `x != nil` because `nil` is not a valid valueref.

peter-b commented on libfoundation/src/foundation-value.cpp:
```diff
@@ -579,7 +640,7 @@ void __MCValueDestroy(__MCValue *self)
 	// a libfoundation function
 #ifdef _DEBUG
-	self -> flags = 0;
+	self -> flags = 0xffffffff;
```
`UINT32_MAX`

peter-b commented on libfoundation/src/foundation-value.cpp:
```diff
@@ -135,9 +141,17 @@ void MCValueRelease(MCValueRef p_value)
+            return;
+        }
+        else
+            ;
```
Could you please remove this empty `else` statement?

peter-b commented on libfoundation/src/foundation-value.cpp:
```diff
@@ -1095,17 +1163,17 @@ void __MCValueFinalize(void)
 #endif /* HAVE_VALGRIND */
 
-			s_value_pools[i] . values = *(__MCValue **)t_value;
+			s_value_pools[i] . values = t_value;
```
This should be:

```
s_value_pools[i].values = t_value->next;
```

There's also a larger problem with `__MCValueFinalize()`. At the moment, the pools are destroyed, and then the unique values, false, true and null constants are destroyed.  However, those are all allocated _from the pools that were just destroyed_, leading to double-frees.  The fix is to move the pool deletion so that it's the last thing that occurs in `__MCValueFinalize()`

peter-b commented:
@livecode-vulcan review not ok 2644249

livecode-vulcan commented:
💔 review by @peter-b not ok 26442499e75eff012d3dbd6f134866ca20ee2d56

runrevmark committed:
[[ Foundation ]] Restructure MCValueRelease

Restructure code paths of MCValueRelease so that it assumes that
non-nil valuerefs are most likely, and reference counts greater than
1 are more likely.

runrevmark committed:
Merge branch 'develop' into foundation-check_valuerefs

runrevmark committed:
[[ Foundation ]] Fix deallocation logic in __MCValueFinalize

Make sure the value pools are the last thing to be deallocated,
otherwise releasing the constant ValueRefs will cause a crash.

peter-b commented:
@livecode-vulcan review ok 4510a9d

livecode-vulcan commented:
💙 review by @peter-b ok 4510a9d71fdfe744769ad7b4a6abe5523eedd405

livecode-vulcan commented:
😎 test success 4510a9d71fdfe744769ad7b4a6abe5523eedd405
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1488)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1507)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1484)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1495)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1507)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1508)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1510)
